### DISCUSSION TITLE
mock gramine's pseudo filesystem for tests?

### DISCUSSION THREAD

**Discussion by nmwael** [gramineproject/gramine#1492] [Q&A] [answered]
So I want to test some parts of my application that relies on gramines pseudo filesystem, like the quote file or the key pair for opening encrypted files. 
Im wondering how other developers have handled this? If you have any ideas please tell me:)

Currently my program are written in RUST, but it could be something else.

**Comment by kailun-qin** [comment, marked as answer]
Hi @nmwael! Thanks for the questions.

> So I want to test some parts of my application that relies on gramines pseudo filesystem, like the quote file or the key pair for opening encrypted files.

Does your parts of application to test rely on the exact behaviors of Gramine file systems (e.g., some Gramine-specific peculiarities/limitations)? If so, I think there is no simple way but to use the real Gramine FS.

However, if your targeting tests rely only on some (contents of) Gramine-SGX specific files (`/dev/attestation/quote` or `/dev/attestation/keys/xxx` as you mentioned), I think you can simply mock them (pls see below).

> Currently my program are written in RUST

What about in your test module, defining a mock type that implements the related traits (e.g., `Read`) w/ the specific structs that the tests expect; also making your functions to test generic over the the trait; then creating/using the mock instances in the test functions?

**Reply by nmwael** [reply]
As no "real" gramine is needed I will look into your suggestion with RUST.. Thanks for the great answer :)
